dialogue-clients/src/main/java/com/palantir/dialogue/clients/PerHostTarget.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.clients;
+
+import com.palantir.dialogue.core.TargetUri;
+
+public final class PerHostTarget {
```

Comment:
We had originally discussed designing `PerHostTarget` to look like:

```java
class PerHostTarget {
    // An opaque hash of the uri and resolvedAddress
    String name;
    boolean isSelf;
}
```

The problem with this approach is that the same logical target may produce different names across Dialogue versions. For example, if we add a new field to `TargetUri` that gets incorporated into the hash. This is bad for a couple reasons:
- Consumers using different Dialogue version will disagree on the sorted order of these channels.
- Consumers may attempt to persist these names and later compare them against names produced by a different Dialogue version.

The only capabilities we require from `PerHostTarget` are that it has a consistent ordering for the same logical target across Dialogue version.

This PR instead accomplishes this by making `TargetUri` implement `Comparable<TargetUri>`. If we add new fields to `TargetUri`, we can implement `compareTo` such that the order remains consistent.

Reply:
I wanted to avoid leaking uri-info to callers, as dialogue aims to provide an abstraction that makes such info unnecessary. However, I don't think it will be possible to provide enough precision for our users without making this data available in some form. `TargetUri` is used to direct requests to individual upstreams (with greater specificity than a URI), so any future additions to abstract away something like nodes behind a reverse proxy (no plans to do this, just speculating about potential future changes) would have to be represented in `TargetUri`, which would be similarly helpful here -- so there shouldn't be any need to expose similar fields from a new type here. I think retaining the `PerHostTarget` wrapper around `TargetUri` buys us some freedom to add new metadata for per-host-channels in the future if needs be later on.

The `isSelf` accessor was an attempt to avoid avoid leaking these details, but I think if we do expose the `TargetUri`, we needn't expose `isSelf` (that sort of check could become library functionality elsewhere, but it isn't required for the dialogue feature imo). The TargetUri will allow consumers to be more accurate in their own `isSelf` check, for instance when N services are running on the same host, the caller is likely to know which port it's listening on, in order to provide further differentiation from other services on the same host.